python-checks/src/main/java/org/sonar/python/checks/tests/SkippedTestNoReasonCheck.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2022 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks.tests;
+
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.ArgList;
+import org.sonar.plugins.python.api.tree.Argument;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Decorator;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.QualifiedExpression;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.StringLiteral;
+import org.sonar.plugins.python.api.tree.Tree;
+
+@Rule(key = "S1607")
+public class SkippedTestNoReasonCheck extends PythonSubscriptionCheck {
+
+  private static final String MESSAGE = "Provide a reason for skipping this test.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.DECORATOR, ctx -> {
+      Decorator decorator = (Decorator) ctx.syntaxNode();
+
+      checkDecoratorSkipWithoutReason(ctx, decorator, "unittest.case.skip");
+      checkDecoratorSkipWithoutReason(ctx, decorator, "pytest.mark.skip");
+    });
+
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, ctx -> {
+      CallExpression callExpression = (CallExpression) ctx.syntaxNode();
+
+      checkCallExpressionSkipWithNoOrEmptyReason(ctx, callExpression, "pytest.skip");
+    });
+  }
+
+  private static void checkDecoratorSkipWithoutReason(SubscriptionContext ctx, Decorator decorator, String decoratorSkipName) {
+    Expression expression = decorator.expression();
+    Symbol symbol = getSymbolFromExpression(expression);
```

Comment:
As of now, it seems we wouldn't do anything if the decorator is a `NAME` (because `getSymbolFromExpression` doesn't handle that case).
I'd suggest to get rid of the `getSymbolFromExpression` method and to replace it here with something like:
```suggestion
    Symbol symbol = null;
    if (expression instanceof HasSymbol) {
      symbol = ((HasSymbol) expression).symbol();
    }
```
That way you can handle any decorator expression which has a symbol with the FQN you seek, without having to care about the underlying tree.

Reply:
This doesn't seems to handle the case of `CallExpression`, which does not implement `HasSymbol` interface. For those, we do not call the method `symbol()` but `calleeSymbol()`.
I still implemented your suggestion in the method `getSymbolFromExpression()` instead of the `QualifiedExpression` check, since as you mentioned it allow to not take care of the tree for this one.